Avoid unnecessary work in the Node destructor

cdumez opened the pull request:
#### 3c3c8d0c2261e71726f571a2b1ebea4a74dfce8e
<pre>
Avoid unnecessary work in the Node destructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=253607">https://bugs.webkit.org/show_bug.cgi?id=253607</a>

Reviewed by Ryosuke Niwa.

Avoid unnecessary work in the Node destructor, as it is hot on Speedometer.
The destructor was calling TextManipulationController::removeNode(), which
would remove the Node from TextManipulationController&apos;s m_manipulatedNodes
and m_textNodesWithNewRenderer. Both of these are WeakHashSet so there is
no need to remove Nodes from these before they get destructed.

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::~Node):

Canonical link: <a href="https://commits.webkit.org/261395@main">https://commits.webkit.org/261395@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d6f8054d6a4852509a4094b8095ff452b10bcd6c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111581 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20715 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/184 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3341 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120339 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115640 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22071 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11799 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3103 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117345 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16396 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/99539 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/36/builds/104446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98345 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/117 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/45205 "layout-tests (failure)") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13202 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/114 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13696 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9559 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19141 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52099 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7935 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15679 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/dom/Node.cpp:
```diff
@@ -424,9 +424,6 @@ Node::~Node()
     ASSERT(!m_previous);
     ASSERT(!m_next);
 
-    if (auto* textManipulationController = document().textManipulationControllerIfExists(); UNLIKELY(textManipulationController))
```
Wouldn't `textManipulationController` be null in most cases anyway?

cdumez replied:
I think so, yes. But it's still an extra branch in the common case that we don't really need anymore because of WeakHashSet.

Constellation replied:
Yeah, and I think this is good. I found that some of performance issue in VMEntryScope comes from the Vector destructor while it is almost always zero sized, and it turned out that Vector destructor bloated the code size of the VMEntryScope destructor. So, if there is a code which is not necessary, removing it would be beneficial for the performance even though it is mostly not taken :)

cdumez committed:
Avoid unnecessary work in the Node destructor
https://bugs.webkit.org/show_bug.cgi?id=253607

Reviewed by Ryosuke Niwa.

Avoid unnecessary work in the Node destructor, as it is hot on Speedometer.
The destructor was calling TextManipulationController::removeNode(), which
would remove the Node from TextManipulationController's m_manipulatedNodes
and m_textNodesWithNewRenderer. Both of these are WeakHashSet so there is
no need to remove Nodes from these before they get destructed.

* Source/WebCore/dom/Node.cpp:
(WebCore::Node::~Node):

Canonical link: https://commits.webkit.org/261395@main
